File: dev/vendor-bootstrap.R
# Refresh inst/lib/bootstrap from an upstream Bootstrap release.
#
# bootstrict ships Bootstrap itself rather than compiling against whatever
# version a theming package happens to bundle, so the version named in the
# DESCRIPTION is the version that reaches the browser. Run this script to move
# to a new release, then update DESCRIPTION and NEWS.md.
#
# Needs node and npm: Bootstrap's prefixes come from autoprefixer, which runs
# after Sass in Bootstrap's own build and has no equivalent in `sass`. See
# vendor-bootstrap.mjs.

version <- "5.3.8"

dest <- file.path(
  "inst",
  "lib",
  "bootstrap"
)
work <- tempfile(
  "bootstrap-vendor"
)
dir.create(
  work
)

tgz <- file.path(
  work,
  "bootstrap.tgz"
)
utils::download.file(
  sprintf(
    "https://registry.npmjs.org/bootstrap/-/bootstrap-%s.tgz",
    version
  ),
  tgz,
  mode = "wb"
)
utils::untar(
  tgz,
  exdir = work
)
src <- file.path(
  work,
  "package"
)

# Bootstrap's own browserslist targets, so the vendored prefixes are the ones
# the official dist/ ships -- the npm tarball leaves .browserslistrc out.
utils::download.file(
  sprintf(
    "https://raw.githubusercontent.com/twbs/bootstrap/v%s/.browserslistrc",
    version
  ),
  file.path(
    work,
    ".browserslistrc"
  )
)

# node resolves imports next to the script, so it runs from `work`, where the
# packages are installed.
scss <- file.path(
  normalizePath(
    dest
  ),
  "scss"
)
file.copy(
  file.path(
    "dev",
    "vendor-bootstrap.mjs"
  ),
  work
)
owd <- setwd(
  work
)
on.exit(
  setwd(
    owd
  ),
  add = TRUE
)
stopifnot(
  system2(
    "npm",
    c(
      "install",
      "--silent",
      "--no-save",
      "postcss",
      "postcss-scss",
      "autoprefixer"
    )
  ) ==
    0
)
unlink(
  scss,
  recursive = TRUE
)
stopifnot(
  system2(
    "node",
    shQuote(c(
      "vendor-bootstrap.mjs",
      file.path(
        src,
        "scss"
      ),
      scss
    ))
  ) ==
    0
)
setwd(
  owd
)

file.copy(
  file.path(
    src,
    c(
      "LICENSE",
      "dist/js/bootstrap.bundle.min.js",
      "dist/js/bootstrap.bundle.min.js.map"
    )
  ),
  dest,
  overwrite = TRUE
)
writeLines(
  version,
  file.path(
    dest,
    "VERSION"
  )
)

File: dev/vendor-bootstrap.mjs
// Bake autoprefixer's output into Bootstrap's .scss sources.
//
// Bootstrap's own build runs autoprefixer *after* Sass. The `sass` R package
// (libsass) has no such step, so the prefixes have to be in the .scss we
// vendor, or the compiled stylesheet is missing what the official dist ships
// (-webkit-user-select, -webkit-mask-*, -webkit-print-color-adjust, ...).
//
// Usage: node vendor-bootstrap.mjs <src scss dir> <dest scss dir>
// Reads browserslist targets from .browserslistrc in the working directory.
import { readdir, readFile, writeFile, mkdir } from 'node:fs/promises'
import { join, dirname, relative } from 'node:path'
import postcss from 'postcss'
import scss from 'postcss-scss'
import autoprefixer from 'autoprefixer'

const [src, dest] = process.argv.slice(2)
const processor = postcss([autoprefixer])

async function scssFiles(dir) {
  const found = []
  for (const entry of await readdir(dir, { withFileTypes: true })) {
    const path = join(dir, entry.name)
    if (entry.isDirectory()) found.push(...(await scssFiles(path)))
    else if (entry.name.endsWith('.scss')) found.push(path)
  }
  return found
}

for (const file of await scssFiles(src)) {
  const to = join(dest, relative(src, file))
  const result = await processor.process(await readFile(file, 'utf8'), {
    from: file,
    to,
    syntax: scss
  })
  await mkdir(dirname(to), { recursive: true })
  await writeFile(to, result.css)
}
